### PR TITLE
Greatest Common Ancestor function; Casper messages use Rholang Par

### PR DESCRIPTION
**Pull request by birchmd**
## Overview
Begin working towards having Casper use blocks that contain real rholang terms. The PR defines methods for detecting when two blocks conflict with each other (so cannot both be in the same canonical history), and includes some other refactoring and some tests.

More PRs coming which will use the Tuplespace to actually do the reductions on newly deployed terms.

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/RHOL-271

### Complete this checklist before you submit the PR
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala [hunk 1]
```diff
@@ -65,6 +65,29 @@ object ProtoUtil {
   def parents(b: BlockMessage): Seq[ByteString] =
     b.header.map(_.parentsHashList).getOrElse(List.empty[ByteString])
 
+  //Two blocks conflict if they both use the same deploy in different histories
+  def conflicts(b1: BlockMessage,
```
[the remark this answers is not in the document]
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala]: Done

**Comment by goral09** [comment]
@birchmd thanks for the test, back to you :)

File: casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala [hunk 2]
```diff
@@ -65,6 +66,29 @@ object ProtoUtil {
...
+                b2: BlockMessage,
+                genesis: BlockMessage,
+                dag: BlockDag): Boolean = {
+    val gca = DagOperations.greatestCommonAncestor(b1, b2, genesis, dag)
+    if (gca == b1 || gca == b2) {
+      //blocks which already in each other's chains do not conflict
```
Comment (pyrocto) [inline casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala]: already *exist* in

**Comment by birchmd** [comment]
In this PR two blocks conflict if the same deploy is included in two different alternate histories. The reason for this definition is that we do not want to include the same deploy multiple times in the canonical history, so a multi-parent block cannot include two conflicting blocks as parents.

A "deploy" itself is uniquely identifiable, even if the same code is deployed at multiple different occasions. This is because a deploy is determined by a user public key, a Rholang term and a nonce, where the nonce gives the uniqueness in the case that the same user wishes to deploy the same code on multiple different occasions.

**Comment by birchmd** [comment]
Note that this does not examine the terms that are being deployed at all. Nor does it look at race winners in the histories, which could be another source of conflict if races were decided differently in different histories and this lead to different outcomes. This more complex form of conflict will need to be handled in future development (probably not until after we are also including the execution trace in blocks as well).

File: casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala [hunk 3]
```diff
@@ -36,4 +41,44 @@ object DagOperations {
           Iterator.empty.next()
         }
     }
+
+  //Conceptually, the GCA is the first point at which the histories of b1 and b2 diverge.
+  //Based on that, we compute by finding the first block from genesis for which there
+  //exists a child of that block which is an ancestor of b1 or b2 but not both.
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/DagOperations.scala]: "ancestor of b1 _and_ b2 but not both."

**Pull request merged**